spec/data-model.md

```diff
@@ -0,0 +1,177 @@
+with a `body` made up of `Text` and `Expression` shapes.
+`Text` represents literal _text_,
+while `Expression` wraps each of the potential _expression_ shapes.
+The `value` of `Text` is the "cooked" value (i.e. escape sequences are processed).
+
+Implementations MUST NOT rely on the set of `Expression` `body` values being exhaustive,
+as future versions of this specification MAY define additional expressions.
+If encountering a `body` with an unrecognised value,
+an implementation SHOULD treat it as it would a `Reserved` value.
+
+```ts
+interface Pattern {
+  body: Array<Text | Expression>
+}
+
+interface Text {
+  type: 'text'
+  value: string
+}
+
+interface Expression {
+  type: 'expression'
+  body: Literal | VariableRef | FunctionRef | Reserved
+}
+```
+
+## Expressions
+
+The `Literal` and `VariableRef` correspond to the the _literal_ and _variable_ syntax rules.
+When they are used as the `body` of an `Expression`,
+they represent _expression_ values with no _annotation_.
+
+An _unquoted_ value is represented by a `Literal` with `quoted: false`,
+while a _quoted_ value would have `quoted: true`.
+The `value` of `Literal` is the "cooked" value (i.e. escape sequences are processed).
+
+In a `VariableRef`, the `name` does not include the initial `$` of the _variable_.
+
+```ts
+interface Literal {
+  type: 'literal'
+  quoted: boolean
+  value: string
+}
+
+interface VariableRef {
+  type: 'variable'
+  name: string
+}
+```
+
+A `FunctionRef` represents an _expression_ with a _function_ _annotation_.
+In a `FunctionRef`,
+the `kind` corresponds to the starting sigil of a _function_:
+`'open'` for `+`, `'close'` for `-`, and `'value'` for `:`.
+The `name` does not include this starting sigil.
+
+If the _expression_ includes a _literal_ or _variable_ before the _annotation_,
+it is included as the `operand`.
+Each _option_ is represented by an `Option`.
+
+```ts
+interface FunctionRef {
+  type: 'function'
+  kind: 'open' | 'close' | 'value'
+  name: string
+  operand?: Literal | VariableRef
+  options?: Option[]
+}
+
+interface Option {
+  name: string
+  value: Literal | VariableRef
+}
+```
+
+A `Reserved` represents an _expression_ with a _reserved_ _annotation_.
+The `sigil` corresponds to the starting sigil of the _reserved_.
+The `source` is the "raw" value (i.e. escape sequences are not processed)
+and includes the starting `sigil`.
+
+Implementations MUST NOT rely on the set of `sigil` values remaining constant,
+as future versions of this specification MAY assign other meanings to such sigils.
+
+If the _expression_ includes a _literal_ or _variable_ before the _annotation_,
+it is included as the `operand`.
+
+```ts
+interface Reserved {
+  type: 'reserved'
+  sigil: '!' | '@' | '#' | '%' | '^' | '&' | '*' | '<' | '>' | '?' | '~'
+  source: string
+  operand?: Literal | VariableRef
+}
+```
+
+## Extensions
+
+Implementations MAY extend this data model with additional fields,
```

Comment:
One reasonable extension point could be adding something like `span: Span` to all interface, with a definition like:

```ts
interface Span {
  start: int
  end: int
}
```

Is there a reason to prevent something like this from being done? Or am I misunderstanding your concern here?

Reply:
Updated as suggested.